wire-ios-data-model/Source/MLS/MLSService.swift

```diff
@@ -1879,61 +1845,84 @@ public final class MLSService: MLSServiceInterface {
+        ///
+        /// Core Crypto can not automatically recover by itself. It needs
+        /// to process incoming handshake messages then generate a new commit.
 
-        } catch CommitError.failedToSendCommit(recovery: .retryAfterQuickSync, cause: let error) {
-            logger.warn("failed to send commit, syncing then retrying operation...")
-            await syncStatus.recoverWithQuickSync()
-            logger.info("sync finished, retying operation...")
+        case retryAfterQuickSync
 
-            guard retryCount <= maxRetryAttempts else {
-                throw error
-            }
+        /// Repair (re-join) the group and retry the action
+        ///
+        /// We may have missed a few commits so we will rejoin the group
+        /// and try again.
 
-            var currentRetryCount = retryCount
-            currentRetryCount += 1
+        case retryAfterRepairingGroup
 
-            try await retryOnCommitFailure(for: groupID, operation: operation, retryCount: currentRetryCount)
+        /// Abort the action and inform the user.
+        ///
+        /// There is no way to automatically recover from the error.
 
-        } catch CommitError.failedToSendCommit(recovery: .retryAfterRepairingGroup, _) {
-            logger.warn("failed to send commit, repairing group then retrying operation...")
-            await fetchAndRepairGroup(with: groupID)
-            logger.info("repair finished, retrying operation...")
-            try await operation()
+        case giveUp
 
-        } catch CommitError.failedToSendCommit(recovery: .giveUp, cause: let error) {
-            logger.warn("failed to send commit, giving up...")
-            throw error
+        init(from reason: String) {
+            if let error = try? MLSAPIError(from: reason) {
+                switch error {
+                case .mlsClientMismatch, .mlsCommitMissingReferences:
+                    self = .retryAfterQuickSync
+                case .mlsStaleMessage:
+                    self = .retryAfterRepairingGroup
+                default:
+                    self = .giveUp
+                }
+            } else {
+                self = .giveUp
+            }
+        }
 
-        } catch ExternalCommitError.failedToSendCommit(recovery: .retry, cause: let error) {
-            logger.warn("failed to send external commit, retrying operation...")
+    }
 
-            guard retryCount <= maxRetryAttempts else {
-                throw error
-            }
+    private func retryOnCommitFailure(
+        for groupID: MLSGroupID,
+        operation: @escaping () async throws -> Void,
+        retryCount: Int = 0
+    ) async throws {
+
+        do {
+            try await operation()
+        } catch let CoreCryptoError.Mls(.MessageRejected(reason: reason)) {
+            switch RecoveryStrategy(from: reason) {
+            case .retryAfterQuickSync:
+                logger.warn("failed to send commit, syncing then retrying operation...")
```

Comment:
suggestion: add groupID to ease debugging
```suggestion
                logger.warn("failed to send commit, syncing then retrying operation...", attributes: [.mlsGroupID: groupID.safeForLoggingDescription])
```